saleor/graphql/discount/mutations/sale/sale_update.py

```diff
@@ -53,57 +63,144 @@ class Meta:
 
     @classmethod
     def perform_mutation(cls, _root, info: ResolveInfo, /, **data):
-        instance = cls.get_instance(info, **data)
-        previous_catalogue = fetch_catalogue_info(instance)
-        previous_end_date = instance.end_date
-        data = data.get("input")
-        manager = get_plugin_manager_promise(info.context).get()
-        cleaned_input = cls.clean_input(info, instance, data)
+        promotion = cls.get_instance(info, **data)
+        input = data.get("input")
+        cls.validate_dates(promotion, input)
+        rules = promotion.rules.all()
+        previous_predicate = rules[0].catalogue_predicate
+        previous_catalogue = convert_migrated_sale_predicate_to_catalogue_info(
+            previous_predicate
+        )
+        previous_end_date = promotion.end_date
+        previous_products = get_products_for_rule(rules[0])
+        previous_product_ids = set(previous_products.values_list("id", flat=True))
         with traced_atomic_transaction():
-            instance = cls.construct_instance(instance, cleaned_input)
-            cls.clean_instance(info, instance)
-            cls.save(info, instance, cleaned_input)
-            cls._save_m2m(info, instance, cleaned_input)
-            current_catalogue = fetch_catalogue_info(instance)
-            cls.send_sale_notifications(
-                manager,
-                instance,
-                cleaned_input,
+            cls.update_fields(promotion, rules, input)
+            cls.clean_instance(info, promotion)
+            promotion.save()
+            for rule in rules:
+                cls.clean_instance(info, rule)
+                rule.save()
+
+            cls.post_save_actions(
+                info,
+                input,
+                promotion,
                 previous_catalogue,
-                current_catalogue,
                 previous_end_date,
+                previous_product_ids,
             )
 
-            cls.update_products_discounted_prices(
-                cleaned_input, previous_catalogue, current_catalogue
-            )
-        return cls.success_response(ChannelContext(node=instance, channel_slug=None))
+        return cls.success_response(ChannelContext(node=promotion, channel_slug=None))
+
+    @classmethod
+    def get_instance(cls, info: ResolveInfo, **data):
+        object_id = cls.get_global_id_or_error(data["id"], "Sale")
+        return models.Promotion.objects.get(old_sale_id=object_id)
```

Comment:
Shouldn't we raise some specific error when the new promotion id is provided, to indicate that user should use the new promotion mutation?